### DISCUSSION TITLE
Optimal Disk/Network Buffer Size

### DISCUSSION THREAD

**Discussion by kgriffs** [falconry/falcon#1819] [General]
This note [from the Python docs](https://docs.python.org/3/library/socket.html#socket.socket.recv) seems to validate the rule of thumb I've used for a long time:

> For best match with hardware and network realities, the value of bufsize should be a relatively small power of 2, for example, 4096.

I wonder what the latest research says on the subject? `io.DEFAULT_BUFFER_SIZE` is currently defined as 8 KiB.

Seems like with buffers getting larger and disks/networks getting faster, you might need to increase the app buffer size to avoid too much context switching. But the reality could very well be different. And you need to be careful not to introduce a lot of buffer bloat.

**Comment by vytas7** [comment]
[Buffer bloat](https://en.wikipedia.org/wiki/Bufferbloat) is a very real issue; excessive buffer bloat may cause huge delays and unexpected jitter (but hey adding more memory cannot be a bad thing? :grimacing: )

Therefore, if in doubt, it's a good idea to stick to the `DEFAULT_BUFFER_SIZE`, or a small multiple of it.
For instance, the ubiquitous Nginx HTTP server also chooses [8 KiB or 16 KiB, depending on the platform architecture](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_body_buffer_size), as its default client body buffer size.

Sadly, pure Python algorithms (particularly poorly written ones! :stuck_out_tongue:) may favour a larger buffer size around 64 KiB due to the interpreter overhead in manipulating small chunks of data (== more ops, more function calls).
